diff --git a/src/puccinialin/_target.py b/src/puccinialin/_target.py
index 8a49763..caef37e 100644
--- a/src/puccinialin/_target.py
+++ b/src/puccinialin/_target.py
@@ -52,7 +52,7 @@ def get_triple(file: typing.IO) -> str:
     shared library of native modules, which should match the binaries that we build on the rust side. For other
     platforms, we may need to query additional fields."""
     soabi = sysconfig.get_config_var("SOABI")
-    if soabi:
+    if soabi and "-" in soabi:
         print(f"Python reports SOABI: {soabi}", file=file)
         if soabi.count("-") == 1:
             _python, sysconfig_platform = soabi.split("-", maxsplit=2)
@@ -80,8 +80,18 @@ def get_triple(file: typing.IO) -> str:
             print(f"Unrecognized linux platform {sysconfig_platform}", file=file)
             sys.exit(1)
         target = f"{arch}-unknown-{linux}-{libc}"
+    # Other Linux
+    elif "linux" in sysconfig_platform:
+        # Check machine architecture to determine the correct target
+        machine = platform.machine()
+        if machine in ("arm64", "aarch64"):
+            target = "aarch64-unknown-linux-gnu"
+        elif machine == "x86_64":
+            target = "x86_64-unknown-linux-gnu"
+        else:
+            raise ValueError(f"Unknown macOS machine: {machine}")
     # macOS (both Intel and arm)
-    elif "darwin" in sysconfig_platform:
+    elif "darwin" in sysconfig_platform or "macosx" in sysconfig_platform:
         # Check machine architecture to determine the correct target
         machine = platform.machine()
         if machine in ("arm64", "aarch64"):
