diff --git a/Modules/objc/libffi_extra.m b/Modules/objc/libffi_extra.m
--- a/Modules/objc/libffi_extra.m
+++ b/Modules/objc/libffi_extra.m
@@ -75,29 +75,11 @@
 {
     ffi_closure* cl;
 
-#ifdef HAVE_CLOSURE_POOL
-
-#if PyObjC_BUILD_RELEASE >= 1015
-    if (@available(macOS 10.15, *)) { // LCOV_BR_EXCL_LINE
-        cl        = ffi_find_closure_for_code_np(codeloc);
-        *cif      = cl->cif;
-        *userdata = cl->user_data;
-        ffi_closure_free(cl);
-    } else
-#endif
-    {
-        cl        = (ffi_closure*)codeloc;
-        *cif      = cl->cif;
-        *userdata = cl->user_data;
-        PyObjC_ffi_closure_free(cl);
-    }
-#else
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wunguarded-availability-new"
-    cl        = ffi_find_closure_for_code_np(codeloc);
+    cl        = (ffi_closure*)codeloc;
     *cif      = cl->cif;
     *userdata = cl->user_data;
     ffi_closure_free(cl);
 #pragma clang diagnostic pop
-#endif
 }
diff --git a/setup.py b/setup.py
--- a/setup.py
+++ b/setup.py
@@ -90,8 +90,6 @@
     "-Wshorten-64-to-32",
     # "-fsanitize=address", "-fsanitize=undefined", "-fno-sanitize=vptr",
     # "--analyze",
-    "-Werror",
-    "-Wno-cast-function-type-mismatch",
     "-I/usr/include/ffi",
     "-fvisibility=hidden",
     "-g",
@@ -683,8 +681,10 @@
             filename = self.get_ext_filename(fullname)
             full_path = os.path.join(self.build_lib, filename)
 
-            if full_path not in prebuild_mtimes or (
-                prebuild_mtimes[full_path] != os.stat(full_path).st_mtime
+            if os.path.exists(full_path) and (
+                full_path not in prebuild_mtimes or (
+                    prebuild_mtimes[full_path] != os.stat(full_path).st_mtime
+                )
             ):
                 subprocess.check_call(["/usr/bin/dsymutil", "-q", full_path])
 
@@ -809,15 +809,6 @@
             extra_compile_args=EXT_CFLAGS,
             extra_link_args=EXT_LDFLAGS,
         ),
-    ]
-    + [
-        Extension(
-            "PyObjCTest." + os.path.splitext(os.path.basename(test_source))[0],
-            [test_source],
-            extra_compile_args=EXT_CFLAGS,
-            extra_link_args=EXT_LDFLAGS,
-        )
-        for test_source in glob.glob(os.path.join("Modules", "objc", "test", "*.[mc]"))
     ],
     cmdclass={
         "build_ext": oc_build_ext,
